test/WithdrawQueue.t.sol

```diff
@@ -0,0 +1,245 @@
+        deal(address(cellar), userC, assetsC);
+
+        {
+            vm.startPrank(userA);
+            WithdrawQueue.WithdrawRequest memory req = WithdrawQueue.WithdrawRequest({
+                deadline: uint64(block.timestamp + 100),
+                inSolve: false,
+                executionSharePrice: 1e6,
+                sharesToWithdraw: uint96(assetsA)
+            });
+            cellar.approve(address(queue), assetsA);
+            queue.updateWithdrawRequest(cellar, req);
+            vm.stopPrank();
+        }
+        {
+            vm.startPrank(userB);
+            WithdrawQueue.WithdrawRequest memory req = WithdrawQueue.WithdrawRequest({
+                deadline: uint64(block.timestamp + 100),
+                inSolve: false,
+                executionSharePrice: 1e6,
+                sharesToWithdraw: uint96(assetsB)
+            });
+            cellar.approve(address(queue), assetsB);
+            queue.updateWithdrawRequest(cellar, req);
+            vm.stopPrank();
+        }
+        {
+            vm.startPrank(userC);
+            WithdrawQueue.WithdrawRequest memory req = WithdrawQueue.WithdrawRequest({
+                deadline: uint64(block.timestamp + 100),
+                inSolve: false,
+                executionSharePrice: 1e6,
+                sharesToWithdraw: uint96(assetsC)
+            });
+            cellar.approve(address(queue), assetsC);
+            queue.updateWithdrawRequest(cellar, req);
+            vm.stopPrank();
+        }
+
+        solverIsCheapskate = true;
+
+        address[] memory users = new address[](3);
+        users[0] = userA;
+        users[1] = userB;
+        users[2] = userC;
+        bytes memory callData = abi.encode(cellar, USDC);
+
+        vm.expectRevert(bytes("TRANSFER_FROM_FAILED"));
+        queue.solve(cellar, users, callData, address(this));
+
+        // But if solver is not a cheapskate.
+        solverIsCheapskate = false;
+
+        // Solve is successful.
+        queue.solve(cellar, users, callData, address(this));
+    }
+
+    // TODO tests for malicious solvers that repeat users, or try to add a user who is not in there,
+    // basically check for all reverts in solve.
+    function testUserRequestWithInSolveTrue() external {
+        WithdrawQueue.WithdrawRequest memory req = WithdrawQueue.WithdrawRequest({
+            deadline: uint64(block.timestamp + 100),
+            inSolve: true,
+            executionSharePrice: 1e6,
+            sharesToWithdraw: 1
+        });
+
+        queue.updateWithdrawRequest(cellar, req);
+
+        WithdrawQueue.WithdrawRequest memory savedReq = queue.getUserWithdrawRequest(address(this), cellar);
+
+        assertTrue(savedReq.inSolve == false, "inSolve should be false");
+    }
+
+    // TODO test showing how user originally wants 100 shares withdrawn, then changes it to 50, make sure only 50 can be withdrawn.
+
+    function finishSolve(bytes calldata runData, uint256, uint256 assetApprovalAmount) external {
```

Comment:
So is this just checking what happens if the solver attempts to solve only half of the amount?

Reply:
So this test had a user submitting a request to withdraw X shares, then they decided to reduce it to be x/2 shares. This test verifies that solvers can only solve x/2 shares.